### PR TITLE
refactor contmoebaseclass

### PR DESCRIPTION
**Pull request by crewtool**
# Description
small MoT code refactor

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/moe_layers/continuous_moe.py [hunk 1]
```diff
@@ -53,19 +53,19 @@ def __post_init__(self):
...
         merge_weights, emit_weights = self.get_merge_and_emit_weights(x)
         x = self.merge_map_emit(x, merge_weights, emit_weights)
         x = self.reshape_into_original(x)
         return x * (self.group_size / self.original_group_size)
 
-    def rearrange_for_grouping(self, x):
+    def reshape_into_groups(self, x):
         """
         :param x: normal input tensor of shape (B, S, dmodel)
         :return: x transposed so that the dimension to group over is second last
```
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:65]: no longer true

**Comment by Simontwice** [comment]
this is some hot stuff you've got there mister
LooksGreatToMe

File: research/conditional/moe_layers/continuous_moe.py (hunk 1 above)
Comment (crewtool) [reply research/conditional/moe_layers/continuous_moe.py:65]: Comments in progress

File: research/conditional/moe_layers/continuous_moe.py [hunk 2]
```diff
@@ -53,32 +53,30 @@ def __post_init__(self):
...
         x = self.merge_map_emit(x, merge_weights, emit_weights)
         x = self.reshape_into_original(x)
         return x * (self.group_size / self.original_group_size)
 
-    def rearrange_for_grouping(self, x):
+    def reshape_into_groups(self, x):
         """
         :param x: normal input tensor of shape (B, S, dmodel)
-        :return: x transposed so that the dimension to group over is second last
+        :return: x of shape (free_dimension, aggr_dimension // group_size, group_size , dmodel)
```
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:65]: explain what free dimension and aggr dimension are

File: research/conditional/moe_layers/continuous_moe.py [hunk 3]
```diff
@@ -97,37 +95,31 @@ def get_temperature(self):
         return self.temperature, self.temperature
 
     def merge_map_emit(self, x, merge_weights, emit_weights):
-        # x shape is free_dimension, aggr_dimension, dmodel
-        # merge_weights shape is free_dimension, aggr_dimension // group_size, group_size, n_experts
+        # merge_weights shape is (free_dimension, aggr_dimension // group_size, group_size, n_experts)
+        # x shape is (free_dimension, aggr_dimension // group_size, group_size, dmodel)
```
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:99]: have x be the first here, also add `merge_weights and emit_weights shape is .,,,`

File: research/conditional/moe_layers/continuous_moe.py [hunk 4]
```diff
@@ -97,37 +95,31 @@ def get_temperature(self):
...
+        # x shape is free_dimension, aggr_dimension // group_size, group_size, dmodel
         return x
 
     def reshape_into_original(self, x):
         if self.sparsity_dim == 0:
             # sequence dimension is the new "batch size" when you think about it
-            x = x.permute(2, 0, 1)
-            # shape is free_dimension, aggr_dimension, dmodel
+            x = x.view(-1, x.size(0), self.dm)
+            # x is reshaped back to B, S, dmodel
```
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:122]: `x is reshaped to the original shape [batch, seq_len, dmodel]`

**Review by Simontwice — CHANGES_REQUESTED** [review]
a

File: research/conditional/moe_layers/continuous_moe.py (hunk 3 above)
Comment (crewtool) [reply research/conditional/moe_layers/continuous_moe.py:99]: Done

File: research/conditional/moe_layers/continuous_moe.py (hunk 2 above)
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:65]: maybe:

Reshape code so the axis to group over is on position 1, and then group the tensor over said axis.
Example: 

- If we group tokens from different sequences in a batch (sparsity = 0), we need to put the batch dimension to position 1.
- If we group tokens within one sequence, the dimension to group over is already on position 1, hence we leave it as is.

File: research/conditional/moe_layers/continuous_moe.py [hunk 5]
```diff
@@ -97,40 +104,38 @@ def get_temperature(self):
...
 
     def reshape_into_original(self, x):
         if self.sparsity_dim == 0:
             # sequence dimension is the new "batch size" when you think about it
-            x = x.permute(2, 0, 1)
-            # shape is free_dimension, aggr_dimension, dmodel
-            return x
+            x = x.view(x.size(0), -1, self.dm)
+            return x.transpose(0, 1)
+            # x is reshaped back to (batch, seq_len, dmodel)
```
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:136]: delete this comment, this function is self explanatory because of its name

Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:136]: delete this comment, the function is self explanatory because of the name

File: research/conditional/moe_layers/continuous_moe.py [hunk 6]
```diff
@@ -97,40 +104,38 @@ def get_temperature(self):
...
+        x = torch.matmul(
+            emit_weights,
+            x.transpose(1, 2),
         )
+        # x shape is free_dimension, split_dimension // group_size, group_size, dmodel
         return x
 
     def reshape_into_original(self, x):
         if self.sparsity_dim == 0:
             # sequence dimension is the new "batch size" when you think about it
```
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:133]: delete this comment

**Review by Simontwice — CHANGES_REQUESTED** [review]
2 things left:

- minor comments 
- speed and memory comparison: before and after PR. Show me 2 experiments on neptune.

File: research/conditional/moe_layers/continuous_moe.py (hunk 5 above)
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:136]: delete comment

File: research/conditional/moe_layers/continuous_moe.py (hunk 6 above)
Comment (Simontwice) [inline research/conditional/moe_layers/continuous_moe.py:133]: delete comment

**Comment by crewtool** [comment]
[Neptune link](https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=99f374ec-39cd-48c4-ac55-0f8f7c35500c&detailsTab=metadata&shortId=LLMRANDOM-2801&dash=leaderboard&type=run&lbViewUnpacked=true&sortBy=%5B%22sys%2Fcreation_time%22%5D&sortFieldType=%5B%22datetime%22%5D&sortFieldAggregationMode=%5B%22auto%22%5D&sortDirection=%5B%22descending%22%5D&query=(%60args%2Fff_mode%60%3Astring%20%3D%20%22cont_moe%22)%20AND%20(%60args%2Fn_experts%60%3Aint%20%3D%20512)%20AND%20(%60args%2Flearning_rate%60%3Afloat%20%3D%200.002)%20AND%20(%60args%2Fgroup_size%60%3Aint%20%3D%2032)%20AND%20(%60args%2Finit_scale%60%3Afloat%20%3D%201)&compare=EwNgnALANAjADAdihCQ)

**Review by Simontwice — APPROVED** [review]
Now we can show people MoT code and expect them to understand, thanks!
